### DISCUSSION TITLE
Any way to use custom logger & formatter for falcon logs

### DISCUSSION THREAD

**Discussion by k1r8r0wn** [socketry/falcon#257] [Q&A]
F.e., I have now this log - `{"time":"2024-10-24T13:19:48+00:00","severity":"info","oid":25200,"pid":1,"subject":"Falcon::Service::Application","message":"Binding to #<Async::HTTP::Endpoint http://0.0.0.0:3000/ {}>..."}` but my custom logger uses another fields...

**Comment by ioquatix** [comment]
Yes you can completely change the log output format.

The documentation is brief but see `CONSOLE_OUTPUT` and https://socketry.github.io/console/guides/integration/index#custom-output-wrapper

**Comment by k1r8r0wn** [comment]
@ioquatix, hey there! Hope you're well!

I try to implement this like this:

.env:
```yaml
CONSOLE_OUTPUT="HappyLogger"
```

lib/console/output/happy_logger.rb:
```ruby
module Console
  module Output
    class HappyLogger
      def self.new(output, **options)
        Rails.logger.info(output, context: options) # I have custom Rails.logger, so actually I need to wrap other events as well (i.e. errors, etc.)
      end
    end
  end
end
```
and ... it worked, but not as I've expected. The reason is that the link with example has another code with `call` method and originally code [here](https://github.com/socketry/console/blob/ebd30eca2797b448a791d2e4149b1f1b56db7207/lib/console/output.rb#L18) calls the self.new as I've understood.

What's more if I set the CONSOLE_OUTPUT="JSON", it doesn't log in json format.

**Reply by ioquatix** [reply]
Here is a full example of how to redirect all Console output to the Rails logger:

```ruby
class RedirectToRailsLogger
  def initialize(output, **options)
    @output = output
    @options = options
  end
  
  def call(*arguments, severity: :info, **options, &block)
    if block_given?
      if block.arity == 0
        arguments << block.call
      else
        buffer = StringIO.new
        block.call(buffer)
        arguments << buffer.string
      end
    end
    
    Rails.logger.public_send(severity, *arguments, **options)
  end
end

ENV["CONSOLE_OUTPUT"] = "RedirectToRailsLogger"

require "rails"
require "console"

Rails.logger = ActiveSupport::TaggedLogging.new(ActiveSupport::Logger.new(STDOUT))

Console.info("Hello, Rails!")
```

If you want JSON output, that would now be a concern of `Rails.logger` to format the output.

**Reply by k1r8r0wn** [reply]
Thank you for the code given, I've checked and method `call` is unreachable.

Also I've still get messages inside the standard logger:

<img width="955" alt="Снимок экрана 2024-11-27 в 15 10 14" src="https://github.com/user-attachments/assets/5511dec0-b2c5-4465-964c-5287a9ab9e52">

P.S. Have earlier the falcon (0.42.3), recently updated to the latest one - falcon (0.48.3).

**Reply by ioquatix** [reply]
If I run the example code, it definitely works:

![image](https://github.com/user-attachments/assets/0af44447-28bb-4de0-9c2d-67bb4755d243)

Can you explain why `call` is unreachable?

**Reply by k1r8r0wn** [reply]
I've meant when I'm running the `falcon host`...
